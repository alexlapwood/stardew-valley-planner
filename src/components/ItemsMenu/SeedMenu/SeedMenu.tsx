import React from "react";

import cn from "classnames";

import BigText from "../../BigText/BigText";
import Sprite from "../../Sprite/Sprite";

import { getSeason, getYear } from "../../../helpers/date";

// tslint:disable-next-line:no-var-requires
const crops: { [index: string]: ICrop } = require("../../../data/sdv.json")
  .crops;

interface IProps {
  date: number;
  isVisible: boolean;
  selectCrop: (cropId: string) => void;
  selectedItem?: ISelectedItem;
}

const SeedMenu: React.SFC<IProps> = props => {
  const { date, isVisible, selectCrop, selectedItem } = props;

  const cropIds = Object.keys(crops).sort((a, b) => a.localeCompare(b));

  const getLastDayForCrop = (crop: ICrop) => {
    let lastDay = 0;

    if (crop.seasons.includes("spring")) {
      lastDay = 28 * 1;
    }

    if (crop.seasons.includes("summer")) {
      lastDay = 28 * 2;
    }

    if (crop.seasons.includes("fall")) {
      lastDay = 28 * 3;
    }

    if (crop.seasons.includes("winter")) {
      lastDay = 28 * 4;
    }

    lastDay += getYear(date) * 28 * 4;

    return lastDay;
  };

  const calculateProfit = (crop: ICrop) => {
    const lastDay = getLastDayForCrop(crop);

    const daysUntilFirstHarvest = crop.stages.reduce(
      (acc, stage) => acc + stage,
      0
    );

    const harvests =
      crop.regrow === undefined
        ? Math.ceil(
            (lastDay - date - daysUntilFirstHarvest) / daysUntilFirstHarvest
          )
        : Math.ceil((lastDay - date - daysUntilFirstHarvest) / crop.regrow);

    return (
      crop.sell * harvests * (crop.harvest.min || 1) -
      crop.buy * (crop.regrow === undefined ? harvests : 1)
    );
  };

  return isVisible ? (
    <div className="SeedMenu sdv-list">
      {cropIds
        .filter(cropId => !crops[cropId].isIndoorsOnly)
        .filter(
          cropId =>
            crops[cropId] &&
            crops[cropId].seasons.find(
              season =>
                season ===
                ["spring", "summer", "fall", "winter"][getSeason(date)]
            )
        )
        .sort((a, b) => a.localeCompare(b))
        .map(cropId => (
          <div
            className={cn("sdv-list-item", {
              selected:
                selectedItem !== undefined &&
                selectedItem.type === "crop" &&
                selectedItem.id === cropId
            })}
            data-automationid={`seed--${cropId}`}
            key={cropId}
            // tslint:disable-next-line:jsx-no-lambda
            onClick={() => {
              selectCrop(cropId);
            }}
            title={
              `Cost: ${crops[cropId].buy}g\n` +
              `Sell: ${crops[cropId].sell}g\n` +
              `Profit: ${calculateProfit(crops[cropId])}g\n` +
              `Profit/day: ${Math.round(
                (calculateProfit(crops[cropId]) /
                  (getLastDayForCrop(crops[cropId]) - date)) *
                  100
              ) / 100}g`
            }
          >
            <div className="sdv-list-item-icon">
              <Sprite
                height={16}
                src="images/seeds.png"
                width={16}
                x={cropIds.indexOf(cropId) * 16}
                y={0}
              />
            </div>
            <div className="sdv-list-item-text">
              <BigText>{crops[cropId].name}</BigText>
            </div>
          </div>
        ))}
    </div>
  ) : null;
};

export default SeedMenu;
