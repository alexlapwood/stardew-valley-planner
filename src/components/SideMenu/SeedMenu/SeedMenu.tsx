import cn from "clsx";
import { For, Show } from "solid-js";

import { crops } from "../../../data/sdv.json";
import { getSeason, getYear } from "../../../helpers/date";
import BigText from "../../BigText/BigText";
import Sprite from "../../Sprite/Sprite";

interface IProps {
  date: number;
  isVisible: boolean;
  selectCrop: (cropId: string) => void;
  selectedItem?: ISelectedItem;
}

export default function SeedMenu(props: IProps) {
  const cropIds = Object.keys(crops).sort((a, b) => a.localeCompare(b));

  const getLastDayOfYearForCrop = (crop: ICrop) => {
    let lastDay = 0;

    if (crop.seasons.includes("spring")) {
      lastDay = 28 * 1;
    }

    if (crop.seasons.includes("summer")) {
      lastDay = 28 * 2;
    }

    if (crop.seasons.includes("fall")) {
      lastDay = 28 * 3;
    }

    if (crop.seasons.includes("winter")) {
      lastDay = 28 * 4;
    }

    lastDay += getYear(props.date) * 28 * 4;

    return lastDay;
  };

  const calculateProfit = (crop: ICrop) => {
    const lastDay = getLastDayOfYearForCrop(crop);

    const daysUntilFirstHarvest = crop.stages.reduce(
      (acc, stage) => acc + stage,
      0
    );

    const harvests =
      crop.regrow === undefined
        ? Math.ceil(
            (lastDay - props.date - daysUntilFirstHarvest) /
              daysUntilFirstHarvest
          )
        : Math.ceil(
            (lastDay - props.date - daysUntilFirstHarvest) / crop.regrow
          );

    return (
      crop.sell * harvests * (crop.harvest.min || 1) -
      crop.buy * (crop.regrow === undefined ? harvests : 1)
    );
  };

  const calculateProfitPerDay = (crop: ICrop) =>
    Math.round(
      (calculateProfit(crop) / (getLastDayOfYearForCrop(crop) - props.date)) *
        100
    ) / 100;

  return (
    <Show when={props.isVisible}>
      <div class="SeedMenu sdv-list">
        <For
          each={cropIds
            .filter((cropId) => !crops[cropId].isIndoorsOnly)
            .filter(
              (cropId) =>
                crops[cropId] &&
                crops[cropId].seasons.find(
                  (season) =>
                    season ===
                    ["spring", "summer", "fall", "winter"][
                      getSeason(props.date)
                    ]
                )
            )
            .sort((a, b) => a.localeCompare(b))}
        >
          {(cropId) => (
            <div
              class={cn("sdv-list-item", {
                selected:
                  props.selectedItem !== undefined &&
                  props.selectedItem.type === "crop" &&
                  props.selectedItem.id === cropId,
              })}
              data-testid={`seed--${cropId}`}
              // tslint:disable-next-line:jsx-no-lambda
              onClick={() => {
                props.selectCrop(cropId);
              }}
              title={
                `Cost: ${crops[cropId].buy}g\n` +
                `Sell: ${crops[cropId].sell}g\n` +
                `Profit: ${calculateProfit(crops[cropId])}g\n` +
                `Profit/day: ${calculateProfitPerDay(crops[cropId])}g`
              }
            >
              <div class="sdv-list-item-icon">
                <Sprite
                  height={16}
                  src="images/seeds.png"
                  width={16}
                  x={cropIds.indexOf(cropId) * 16}
                  y={0}
                />
              </div>
              <div class="sdv-list-item-text">
                <BigText>{crops[cropId].name}</BigText>
              </div>
            </div>
          )}
        </For>
      </div>
    </Show>
  );
}
