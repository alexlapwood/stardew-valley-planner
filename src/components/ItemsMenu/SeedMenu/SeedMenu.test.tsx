import { mount } from "enzyme";
import React from "react";

import SeedMenu from "./SeedMenu";

describe("<SeedMenu />", () => {
  it("updates the selected plant when a plant is clicked", () => {
    const selectCropMock = jest.fn();

    const itemsMenu = mount(
      <SeedMenu date={0} isVisible={true} selectCrop={selectCropMock} />
    );

    itemsMenu.find("[data-automationid='seed--green_bean']").simulate("click");

    expect(selectCropMock).toHaveBeenCalledWith("green_bean");
  });

  describe("Crops that don't regrow", () => {
    it("calculates the profit correctly", () => {
      const itemsMenu = mount(
        <SeedMenu date={0} isVisible={true} selectCrop={jest.fn()} />
      );

      const tooltip = itemsMenu
        .find("[data-automationid='seed--parsnip']")
        .prop("title");

      expect(tooltip).toMatchSnapshot();
    });

    it("calculates the profit correctly for year 2", () => {
      const itemsMenu = mount(
        <SeedMenu date={28 * 4} isVisible={true} selectCrop={jest.fn()} />
      );

      const tooltip = itemsMenu
        .find("[data-automationid='seed--parsnip']")
        .prop("title");

      expect(tooltip).toMatchSnapshot();
    });

    it("calculates the profit correctly for crops planted just before they go out of season", () => {
      const itemsMenu = mount(
        <SeedMenu date={23} isVisible={true} selectCrop={jest.fn()} />
      );

      const tooltip = itemsMenu
        .find("[data-automationid='seed--parsnip']")
        .prop("title");

      expect(tooltip).toMatchSnapshot();
    });

    it("calculates the profit correctly for crops planted out of season", () => {
      const itemsMenu = mount(
        <SeedMenu date={24} isVisible={true} selectCrop={jest.fn()} />
      );

      const tooltip = itemsMenu
        .find("[data-automationid='seed--parsnip']")
        .prop("title");

      expect(tooltip).toMatchSnapshot();
    });
  });

  describe("Crops that don regrow", () => {
    it("calculates the profit correctly", () => {
      const itemsMenu = mount(
        <SeedMenu date={0} isVisible={true} selectCrop={jest.fn()} />
      );

      const tooltip = itemsMenu
        .find("[data-automationid='seed--green_bean']")
        .prop("title");

      expect(tooltip).toMatchSnapshot();
    });

    it("calculates the profit correctly for crops planted just before they go out of season", () => {
      const itemsMenu = mount(
        <SeedMenu date={2} isVisible={true} selectCrop={jest.fn()} />
      );

      const tooltip = itemsMenu
        .find("[data-automationid='seed--green_bean']")
        .prop("title");

      expect(tooltip).toMatchSnapshot();
    });

    it("calculates the profit correctly for crops planted out of season", () => {
      const itemsMenu = mount(
        <SeedMenu date={3} isVisible={true} selectCrop={jest.fn()} />
      );

      const tooltip = itemsMenu
        .find("[data-automationid='seed--green_bean']")
        .prop("title");

      expect(tooltip).toMatchSnapshot();
    });
  });
});
