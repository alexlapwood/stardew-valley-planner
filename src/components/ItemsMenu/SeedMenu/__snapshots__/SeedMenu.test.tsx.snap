// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SeedMenu /> Crops that don regrow calculates the profit correctly 1`] = `
"Cost: 60g
Sell: 40g
Profit: 180g
Profit/day: 6.43g"
`;

exports[`<SeedMenu /> Crops that don regrow calculates the profit correctly for crops planted just before they go out of season 1`] = `
"Cost: 60g
Sell: 40g
Profit: 180g
Profit/day: 6.92g"
`;

exports[`<SeedMenu /> Crops that don regrow calculates the profit correctly for crops planted out of season 1`] = `
"Cost: 60g
Sell: 40g
Profit: 140g
Profit/day: 5.6g"
`;

exports[`<SeedMenu /> Crops that don't regrow calculates the profit correctly 1`] = `
"Cost: 20g
Sell: 35g
Profit: 90g
Profit/day: 3.21g"
`;

exports[`<SeedMenu /> Crops that don't regrow calculates the profit correctly for crops planted just before they go out of season 1`] = `
"Cost: 20g
Sell: 35g
Profit: 15g
Profit/day: 3g"
`;

exports[`<SeedMenu /> Crops that don't regrow calculates the profit correctly for crops planted out of season 1`] = `
"Cost: 20g
Sell: 35g
Profit: 0g
Profit/day: 0g"
`;

exports[`<SeedMenu /> Crops that don't regrow calculates the profit correctly for year 2 1`] = `
"Cost: 20g
Sell: 35g
Profit: 90g
Profit/day: 3.21g"
`;
